### PR TITLE
Part1

### PR DESCRIPTION
**Pull request by dpsproservices**
Ready for pull request review.

### CODE REVIEW

File: main.cpp [hunk 1]
```diff
@@ -100,18 +106,45 @@ int main()
...
+    for( Motorcycle& motorcycle : motorcycles ) 
+    {
+        highway.addVehicle(&motorcycle);
+    }
+
+    //  8) Add some SemiTrucks to the highway.
+    for( SemiTruck& semitruck : trucks ) 
+    {
+        highway.addVehicle(&semitruck);
+    }
```
Comment (matkatmusic) [inline main.cpp]: You can save some time by writing `for( auto& element : container )`.  If you change the type held by the container, you won't get error messages as you would currently with what is written.
- [ ] revise if you want.

File: SemiTruck.h [hunk 2]
```diff
@@ -1,2 +1,14 @@
...
+
+struct SemiTruck : public Vehicle
+{
+    SemiTruck(const std::string& n);
+
+    ~SemiTruck() override;
+    SemiTruck(const SemiTruck&);
+    SemiTruck& operator = (const SemiTruck&);
+
+    void pullOver();
```
Comment (matkatmusic) [inline SemiTruck.h:13]: WHere's the function to call when they enter the highway? 
- [ ] revise

File: HighwayPatrol.cpp [hunk 3]
```diff
@@ -1,16 +1,20 @@
...
+HighwayPatrol::~HighwayPatrol() = default;
+HighwayPatrol::HighwayPatrol (const HighwayPatrol&) = default;
+HighwayPatrol& HighwayPatrol::operator = (const HighwayPatrol&) = default;  
 
 void HighwayPatrol::scanHighway(Highway* h)
 {
     std::cout << name << ": scanning highway for speeders" << std::endl;
 
-    for( size_t i = h->vehicles.size(); --i >= 0; )
+    for( size_t i = h->vehicles.size() - 1; static_cast<int>(i) >= 0; --i)
```
Comment (matkatmusic) [inline HighwayPatrol.cpp]: While this works, what I had after fixing the bug another person spotted here (which didn't get pushed to your project) is this:
```
for( size_t i = h->vehicles.size(); i-- > 0; )
```
Take a look here: http://coliru.stacked-crooked.com/a/1c450220662e32f8 
- [ ] revise to eliminate that `static_cast<>`

File: Highway.cpp [hunk 4]
```diff
@@ -13,22 +14,55 @@ void Highway::changeSpeed(int newSpeed)
...
+        car->closeWindows();
+    }
+    else if( auto* motorcycle = dynamic_cast<Motorcycle*>(v) )
+    {
+        motorcycle->lanesplitAndRace( 180 );
+    }
+    else if( auto* semitruck = dynamic_cast<SemiTruck*>(v) )
+    {
+        semitruck->pullOver();
+    }
```
Comment (matkatmusic) [inline Highway.cpp]: Really, you want the semi-truck to pull over when it enters the highway?
- [ ] revise.

File: Car.cpp [hunk 5]
```diff
@@ -14,4 +18,9 @@ void Car::tryToEvade()
...
     setSpeed(0);
     std::cout << name << ": hello officer, what seems to be the problem?" << std::endl;
-}
\ No newline at end of file
+}
+
+void Car::setSpeed(int s)
+{
+    Vehicle::setSpeed(s);
+}
```
Comment (matkatmusic) [inline Car.cpp]: You have not added any extra functionality to the base class implementation.  remove this function.
- [ ] revise

**Review by matkatmusic — CHANGES_REQUESTED** [review]
Close.  needs some tweaking.

File: Car.cpp (hunk 5 above)
Comment (dpsproservices) [reply Car.cpp]: done.

File: HighwayPatrol.cpp (hunk 3 above)
Comment (dpsproservices) [reply HighwayPatrol.cpp]: Cool I had tried something similar but figured you would not want i--  haha.

File: main.cpp (hunk 1 above)
Comment (dpsproservices) [reply main.cpp]: Cool ok done.

File: Highway.cpp (hunk 4 above)
Comment (dpsproservices) [reply Highway.cpp]: Added SemiTruck::cruiseInRightLane()

File: SemiTruck.h (hunk 2 above)
Comment (dpsproservices) [reply SemiTruck.h:13]: Added SemiTruck::cruiseInRightLane()

**Review by matkatmusic — APPROVED** [review]

**Comment by dpsproservices** [comment]
Conflicts resolved

**Review by matkatmusic — APPROVED** [review]

**Pull request merged**